File: ios/Sources/AppAttestPlugin/AppAttestPlugin.swift
import Foundation
import Capacitor
import DeviceCheck

@objc(AppAttestPlugin)
public class AppAttestPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AppAttestPlugin"
    public let jsName = "AppAttest"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isSupported", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "generateKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "attestKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "generateAssertion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "storeKeyId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getStoredKeyId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearStoredKeyId", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = AppAttest()

    @objc func isSupported(_ call: CAPPluginCall) {
        call.resolve(["isSupported": implementation.isSupported()])
    }

    @objc func generateKey(_ call: CAPPluginCall) {
        do {
            try validateSupport()

            DCAppAttestService.shared.generateKey { [weak self] keyId, error in
                guard let self else { return }

                if let error {
                    self.reject(call, message: "Error generating key: \(error.localizedDescription)", error: error)
                    return
                }

                guard let keyId else {
                    self.reject(call, message: AppAttestError.missingGeneratedValue.localizedDescription)
                    return
                }

                self.resolve(call, payload: ["keyId": keyId])
            }
        } catch {
            reject(call, message: error.localizedDescription, error: error)
        }
    }

    @objc func attestKey(_ call: CAPPluginCall) {
        do {
            try validateSupport()
            let keyId = try keyId(from: call)
            let challenge = try challenge(from: call)
            let clientDataHash = try implementation.createClientDataHash(from: challenge)

            DCAppAttestService.shared.attestKey(keyId, clientDataHash: clientDataHash) { [weak self] attestation, error in
                guard let self else { return }

                if let error {
                    self.reject(call, message: "Error attesting key: \(error.localizedDescription)", error: error)
                    return
                }

                guard let attestation else {
                    self.reject(call, message: AppAttestError.missingGeneratedValue.localizedDescription)
                    return
                }

                self.resolve(call, payload: [
                    "attestation": attestation.base64EncodedString(),
                    "keyId": keyId,
                    "challenge": challenge
                ])
            }
        } catch {
            reject(call, message: error.localizedDescription, error: error)
        }
    }

    @objc func generateAssertion(_ call: CAPPluginCall) {
        do {
            try validateSupport()
            let keyId = try keyId(from: call)
            let payload = try payload(from: call)
            let clientDataHash = try implementation.createClientDataHash(from: payload)

            DCAppAttestService.shared.generateAssertion(keyId, clientDataHash: clientDataHash) { [weak self] assertion, error in
                guard let self else { return }

                if let error {
                    self.reject(call, message: "Error generating assertion: \(error.localizedDescription)", error: error)
                    return
                }

                guard let assertion else {
                    self.reject(call, message: AppAttestError.missingGeneratedValue.localizedDescription)
                    return
                }

                self.resolve(call, payload: [
                    "assertion": assertion.base64EncodedString(),
                    "keyId": keyId
                ])
            }
        } catch {
            reject(call, message: error.localizedDescription, error: error)
        }
    }

    @objc func storeKeyId(_ call: CAPPluginCall) {
        do {
            let keyId = try keyId(from: call)
            implementation.storeKeyId(keyId)
            call.resolve(["success": true])
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func getStoredKeyId(_ call: CAPPluginCall) {
        let keyId = implementation.getStoredKeyId()
        call.resolve([
            "keyId": keyId as Any,
            "hasStoredKey": keyId != nil
        ])
    }

    @objc func clearStoredKeyId(_ call: CAPPluginCall) {
        implementation.clearStoredKeyId()
        call.resolve(["success": true])
    }

    private func validateSupport() throws {
        guard implementation.isSupported() else {
            throw AppAttestError.notSupported
        }
    }

    private func keyId(from call: CAPPluginCall) throws -> String {
        guard let keyId = call.getString("keyId"), !keyId.isEmpty else {
            throw AppAttestError.missingKeyId
        }
        return keyId
    }

    private func challenge(from call: CAPPluginCall) throws -> String {
        guard let challenge = call.getString("challenge"), !challenge.isEmpty else {
            throw AppAttestError.missingChallenge
        }
        return challenge
    }

    private func payload(from call: CAPPluginCall) throws -> String {
        guard let payload = call.getString("payload"), !payload.isEmpty else {
            throw AppAttestError.missingPayload
        }
        return payload
    }

    private func resolve(_ call: CAPPluginCall, payload: [String: Any]) {
        DispatchQueue.main.async {
            call.resolve(payload)
        }
    }

    private func reject(_ call: CAPPluginCall, message: String, error: Error? = nil) {
        DispatchQueue.main.async {
            call.reject(message, nil, error)
        }
    }
}

File: android/src/main/java/app/capgo/plugin/appattest/AppAttestPlugin.java
package app.capgo.plugin.appattest;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.android.play.core.integrity.IntegrityManagerFactory;
import com.google.android.play.core.integrity.StandardIntegrityException;
import com.google.android.play.core.integrity.StandardIntegrityManager;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@CapacitorPlugin(name = "AppAttest")
public class AppAttestPlugin extends Plugin {

    private static final String INTEGRITY_ERROR = "INTEGRITY_ERROR";
    private static final String CONFIG_CLOUD_PROJECT_NUMBER = "cloudProjectNumber";
    private static final String DEFAULT_ANDROID_KEY_ID = "android-standard-integrity";

    private final AppAttest implementation = new AppAttest();
    private final Map<String, StandardIntegrityManager.StandardIntegrityTokenProvider> tokenProviders = new ConcurrentHashMap<>();

    @PluginMethod
    public void isSupported(PluginCall call) {
        JSObject response = new JSObject();
        response.put("isSupported", implementation.isSupported(getContext()));
        call.resolve(response);
    }

    @PluginMethod
    public void generateKey(PluginCall call) {
        if (!implementation.isSupported(getContext())) {
            call.reject("Play Integrity is not supported on this device");
            return;
        }

        prepareProvider(call, DEFAULT_ANDROID_KEY_ID, (provider) -> {
            JSObject response = new JSObject();
            response.put("keyId", DEFAULT_ANDROID_KEY_ID);
            call.resolve(response);
        });
    }

    @PluginMethod
    public void attestKey(PluginCall call) {
        String keyId = call.getString("keyId");
        String challenge = call.getString("challenge");

        if (isBlank(keyId)) {
            call.reject("keyId is required");
            return;
        }

        if (isBlank(challenge)) {
            call.reject("challenge is required");
            return;
        }

        requestStandardIntegrityToken(call, keyId, challenge, "attestation", true);
    }

    @PluginMethod
    public void generateAssertion(PluginCall call) {
        String keyId = call.getString("keyId");
        String payload = call.getString("payload");

        if (isBlank(keyId)) {
            call.reject("keyId is required");
            return;
        }

        if (isBlank(payload)) {
            call.reject("payload is required");
            return;
        }

        requestStandardIntegrityToken(call, keyId, payload, "assertion", false);
    }

    @PluginMethod
    public void storeKeyId(PluginCall call) {
        String keyId = call.getString("keyId");
        if (isBlank(keyId)) {
            call.reject("keyId is required");
            return;
        }

        prepareProvider(call, keyId, (provider) -> {
            JSObject response = new JSObject();
            response.put("success", true);
            call.resolve(response);
        });
    }

    @PluginMethod
    public void getStoredKeyId(PluginCall call) {
        String keyId = null;
        if (tokenProviders.containsKey(DEFAULT_ANDROID_KEY_ID)) {
            keyId = DEFAULT_ANDROID_KEY_ID;
        } else if (!tokenProviders.isEmpty()) {
            keyId = tokenProviders.keySet().iterator().next();
        }

        JSObject response = new JSObject();
        response.put("keyId", keyId);
        response.put("hasStoredKey", keyId != null);
        call.resolve(response);
    }

    @PluginMethod
    public void clearStoredKeyId(PluginCall call) {
        tokenProviders.clear();
        JSObject response = new JSObject();
        response.put("success", true);
        call.resolve(response);
    }

    private void requestStandardIntegrityToken(
        PluginCall call,
        String keyId,
        String payload,
        String responseField,
        boolean includeChallenge
    ) {
        if (!implementation.isSupported(getContext())) {
            call.reject("Play Integrity is not supported on this device");
            return;
        }

        final String requestHash;
        try {
            requestHash = implementation.createRequestHash(payload);
        } catch (Exception error) {
            call.reject("Failed to generate request hash", INTEGRITY_ERROR, error);
            return;
        }

        withProvider(call, keyId, (provider) -> {
            StandardIntegrityManager.StandardIntegrityTokenRequest tokenRequest =
                StandardIntegrityManager.StandardIntegrityTokenRequest.builder().setRequestHash(requestHash).build();

            provider
                .request(tokenRequest)
                .addOnSuccessListener((token) -> {
                    JSObject result = new JSObject();
                    result.put(responseField, token.token());
                    result.put("keyId", keyId);
                    if (includeChallenge) {
                        result.put("challenge", payload);
                    }
                    call.resolve(result);
                })
                .addOnFailureListener((error) -> rejectPlayIntegrityError(call, "Play Integrity token request failed", error));
        });
    }

    private void withProvider(PluginCall call, String keyId, ProviderReadyCallback callback) {
        StandardIntegrityManager.StandardIntegrityTokenProvider existingProvider = tokenProviders.get(keyId);
        if (existingProvider != null) {
            callback.onReady(existingProvider);
            return;
        }

        if (DEFAULT_ANDROID_KEY_ID.equals(keyId)) {
            prepareProvider(call, keyId, callback);
            return;
        }

        call.reject("Unknown Android keyId. Call generateKey() or storeKeyId() first to prepare a native Play Integrity provider.");
    }

    private void prepareProvider(PluginCall call, String keyId, ProviderReadyCallback callback) {
        final long cloudProjectNumber;
        try {
            cloudProjectNumber = resolveCloudProjectNumber(call);
        } catch (IllegalArgumentException error) {
            call.reject(error.getMessage());
            return;
        }

        StandardIntegrityManager integrityManager = IntegrityManagerFactory.createStandard(getContext());
        StandardIntegrityManager.PrepareIntegrityTokenRequest prepareRequest =
            StandardIntegrityManager.PrepareIntegrityTokenRequest.builder().setCloudProjectNumber(cloudProjectNumber).build();

        integrityManager
            .prepareIntegrityToken(prepareRequest)
            .addOnSuccessListener((provider) -> {
                tokenProviders.put(keyId, provider);
                callback.onReady(provider);
            })
            .addOnFailureListener((error) -> rejectPlayIntegrityError(call, "Failed to prepare Play Integrity token request", error));
    }

    private long resolveCloudProjectNumber(PluginCall call) {
        Long callValueLong = call.getLong(CONFIG_CLOUD_PROJECT_NUMBER);
        if (callValueLong != null) {
            return callValueLong;
        }

        String callValueString = call.getString(CONFIG_CLOUD_PROJECT_NUMBER);
        if (!isBlank(callValueString)) {
            return parseCloudProjectNumber(callValueString);
        }

        String configValueString = getConfig().getString(CONFIG_CLOUD_PROJECT_NUMBER);
        if (!isBlank(configValueString)) {
            return parseCloudProjectNumber(configValueString);
        }

        throw new IllegalArgumentException(
            "cloudProjectNumber is required on Android. Set AppAttest.cloudProjectNumber in capacitor config or pass it in options."
        );
    }

    private long parseCloudProjectNumber(String value) {
        try {
            return Long.parseLong(value);
        } catch (NumberFormatException error) {
            throw new IllegalArgumentException("cloudProjectNumber must be a valid integer string");
        }
    }

    private boolean isBlank(String value) {
        return value == null || value.trim().isEmpty();
    }

    private void rejectPlayIntegrityError(PluginCall call, String message, Exception error) {
        if (error instanceof StandardIntegrityException) {
            StandardIntegrityException integrityError = (StandardIntegrityException) error;
            call.reject(
                message + " with code " + integrityError.getErrorCode() + ": " + integrityError.getMessage(),
                INTEGRITY_ERROR,
                error
            );
            return;
        }
        call.reject(message, INTEGRITY_ERROR, error);
    }

    @FunctionalInterface
    private interface ProviderReadyCallback {
        void onReady(StandardIntegrityManager.StandardIntegrityTokenProvider provider);
    }
}
